README
dyncall - A Generic Dynamic FFI package
=======================================

dyncall is a low-level toolkit providing a portable abstraction for 
handling native code dynamically at run time.
It comprises three independent components, available as C libraries, namely:

 - 'dyncall' for making function calls,
 - 'dyncallback' or writing generic callback handlers, and
 - 'dynload' for loading code.

The package provides abstractions to the Application Binary Interface 
of various hardware platforms such as x86, AMD64, ARM32/64, PowerPCC32/64,
MIPS32/64 and SPARC32/64.

Our releases are thoroughly tested across all supported platforms using a
flexible test suite.

A typical application for the use of dyncall is the back-end of a FFI 
(foreign function interface) for scripting languages to connect with 
the C/C++ world.

The binary footprint of the package is approx. 10-20 kb, depending on
the platform.
Besides the test suite already mentioned, comprehensive documentation is
available.


Building the libraries
----------------------

Prequisites: 
 - C compiler such as gcc, clang, sunpro or msvc.
 - Assembler such as gas (gnu/apple), fbe or masm.
 - Build system such as 'shell/make', cmake, nmake (windows) or mk (plan9).


Supported Architectures and ABIs:
---------------------------------

 - x86     32/64-bit  (cdecl, fastcall, stdcall, C++ this calls, syscalls, plan9)
 - ARM     AArch32/64 (32: ARM/Thumb mode, v4t-v7, oabi/eabi, soft/hardfloat ABIs; 64: AAPCS64)
 - MIPS    32/64-bit  (o32, n64, eabi)
 - PowerPC 32/64-bit  (32: Mac OS X, System V; 64: System V little and big endian ELF ABI)
 - Sparc   32/64-bit  (v7/8 and v9)

NOTE: Callback support is currently available on
      x86-32/64, arm32 softfloat ABIs, arm64, PPC64 and PPC32 Linux/OS X.

Supported Operating Systems:
----------------------------

 - Android
 - *BSD
 - Haiku
 - iOS
 - Linux
 - Mac OS X
 - Microsoft Windows
 - Minix
 - Nintendo DS Homebrew
 - Plan9 / 9front
 - Playstation Portable Homebrew
 - ReactOS
 - Solaris


Building the libraries
----------------------

Using configure shell script and GNU/BSD/Sun make system.

$ ./configure
$ make
$ make install

Testing:

$ make tests
$ make run-tests


Build 'out-of-source'
---------------------

$ cd <build_dir>
$ <src_dir>/configure --prefix=<install_dir>
$ make
$ sudo make install


Building using CMake
--------------------

$ cmake 
$ make
$ sudo make install


Building on Windows
-------------------

See doc/README.Windows


Platform-specific build details
-------------------------------

See doc/README.*


Building the manual
-------------------

Make sure pdflatex is installed and run

$ make docs


Please refer to http://www.dyncall.org, dyncall(3), dyncallback(3), the dyncall
manual and platform-specific READMEs under doc/ for more information.



ChangeLog
This file lists bigger/noteworthy changes, only...

Version 1.0 (upcoming)
dyncall:
  o PPC64 single-precision float fixes for more than 13 float args (thanks Masanori!)
  o fixed potential buffer overrun on ARM64
  o error code handling (dcGetError) changes, flag is now cleared (work still
    underway)
  o MIPS n64 ellipsis call fixes
  o ARM32 armhf ellipsis call fixes
  o better C++ compiler support (for pointer conversions)
dyncallback:
  o PPC64 single-precision float fixes for more than 13 float args (thanks Masanori!)
  o MIPS o32 (big- and little-endian), EABI (32 bit, little-endian) and n64 (big-endian) support
doc:
  o working html doc generation from TEX sources
bindings:
  o better documentation, removed relative path dependencies, general cleanup
  o python: updated to latest signature format (was still on 0.1)
  o java: rewritten (old version was never complete, anyways)

Version 0.9 (2015/12/31)
dyncall:
  o ARM64 support (AAPCS64 on Linux & Apple's derived version on iOS)
  o armhf calling convention support for ARM32
  o PPC64 support, including syscalls (thanks Masanori!)
  o introduced dcArgF and dcVArgF for flexible/convenient argument binding (works like dcCallF/dcVCallF but without the function call)
  o using -fPIC by default for FreeBSD
  o PPC32 linux syscall support
  o fixed return value errors for ARM32 on some platforms (e.g. freebsd/arm32, nintendo ds); arm & thumb mode
dyncallback:
  o ARM64 support (AAPCS64 on Linux, Apple's derived version on iOS)
  o armhf calling convention support for ARM32
  o PPC64 support (thanks Masanori!)
  o PPC32 System V support (Linux)
  o fix for PPC32 Darwin/OS X
  o more intuitive specification of callback return type
doc:
  o added ARM64 support info and calling convention documentation
  o added ARM32 armhf section
  o added PPC64 support info
  o cleanup, brought all up to date, etc.
tests:
  o callf test covers new argf interface
  o several fixes to some test suites
buildsys:
  o more flexible testsuite build and 'pack' target for convenience
  o fixes/additions for OS X universal builds
  o added iOS universal build support for armv7/arm64
  o helper script for generic .elf -> EBOOT.PBP creation for PSP
  o Sun make fixes for C++ code in test suite
bindings:
  o erldc: Erlang binding (thanks Erik!)
infrastructure:
  o moved to hg
  o moved bindings to own repository

Version 0.8 (2014/03/24)
buildsys:
  o big simplification make-based build system, no more distinct GNU and BSD makefiles anymore
  o new, much simpler ./configure for new make-based build-system
  o support for in-source and out-of-source builds for make-based build systems
  o made configure.bat guess x64 as target arch for x64 windows hosts
  o based more sources on portasm generation
dyncall:
  o fix for ARMv7 Thumb-2 code (tested on iPhone 4 CDMA / iOS 6.0.1)
  o bugfixes for -O2 builds for SPARC-v9 and SPARC-v7
  o new optimized call kernel for SPARC-v9 without conditionals
  o bigger refactoring and optimization of MIPS o32
  o x64 optimizations
dyncallback:
  o added callback support for plan9
  o big cleanup, removal of duplicate files after introduction of portasm
  o include path fixes for public facing APIs (removed relative paths)
doc:
  o cleanup of manual and corrections
  o added info about newly supported and/or tested platforms (e.g. FreeBSD on ARM)
  o restructured folder layout for documentation
  o added API doc for dyncallback and dynload
  o added manpages for dyncallback and dynload
tests:
  o added arm/thumb interwork test suite
bindings:
  o godc: added bindings for Go
  o rbdc: fixed and extended ruby bindings
  o shdc: renamed shell binding to shdc and added feature to list shared-objects' symbols

Version 0.7 (2012/01/02)
architecture:
  o added subproject: portasm - portable (GNU AS,+Apple and MASM) assembly framework
  o added subproject: autovar - predefined macro framework (was dyncall_macros.h)
  o added support for sparc (32-bit) (tested on linux/debian,openbsd,solaris)
  o added support for sparc (64-bit) (tested on linux/debian,openbsd,solaris)
  o added support for Solaris and SunPro compiler for i386, x86_64, sparc and sparc64
  o improved auto-detection via preprocessor defines
dyncall updates:
  o uses portasm (GNU,Apple,Microsoft) common sources for x86,x64,ppc32,arm32_arm
  o bug fixes for x64/win64 calling convention
  o added support for mingw32-w64 on x64
  o improved cdecl x86 calling convention: stack always 16 byte aligned
  o updated API and initial support for structs (for x86/AMD64)
  o initial support for syscalls on x86, added x86 int80h linux/bsd
  o added Minix port (stable)
  o optimized vector buffer for small type pushing
  o minor fix for amd64/sysv: alignment to 32-byte
  o bug fix for ppc32/sysv: ellipsis calls work now
  o updated API, introduced DC_CALL_C_ELLIPSIS_VARARGS
  o changed interface: dcMode does not reset internally anymore
dyncallback updates:
  o uses portasm for x86,x64
  o added Solaris/x86/sunpro port (stable)
  o added Minix port (unstable)
  o callback support for Darwin/ppc32
  o updates source, uses relative includes
dynload updates:
  o fixes for missing includes (dyncall_alloc.h)
  o fixes for self-resolving symbols on windows
  o removed Dbghelf.lib dependency
  o bug fixes for cygwin/x86
  o fixes for beos/haiku
  o no dependencies to libdyncall sources.
buildsys additions:
  o added zero-config (BSD,GNU,SUN) make files (Makefile.embedded)
  o added in/out-of-source configure2 w/ (BSD,GNU) make files (Makefile.generic)
  o added bootstrap lua script (download/build)
  o Nmakefile: more tests included (resolve_self)
  o improved cross-compilation for iOS (upgrade to 4.3 sdk)
  o darwin 8.0 support
  o added 'install' support for Makefile.embedded
buildsys/gmake updates:
  o added support for msvc/x64 tool-chain
  o default settings for build-dir changed to '.'
  o cleanup: removed top-level Make{Prolog,Epilog,Rules} files and updated all sub-projects and tests

dyncall/dyncall_callvm_ppc64.c
/*

 Package: dyncall
 Library: dyncall
 File: dyncall/dyncall_callvm_ppc64.c
 Description: 
 License:

   Copyright (c) 2014-2015 Masanori Mitsugi <mitsugi@linux.vnet.ibm.com>

   Permission to use, copy, modify, and distribute this software for any
   purpose with or without fee is hereby granted, provided that the above
   copyright notice and this permission notice appear in all copies.

   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
   WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
   MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
   ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
   WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
   ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
   OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

*/


/*

  dyncall callvm for ppc64 architectures

  SUPPORTED CALLING CONVENTIONS
  ppc64/linux
  ppc64/syscall

  REVISION
  2015/07/08 added syscall
  2014/08/07 initial support

*/

#include "dyncall_callvm_ppc64.h"
#include "dyncall_call_ppc64.h"
#include "dyncall_alloc.h"
#include "dyncall_macros.h"
#include "dyncall_types.h"

/* Support for Power PC 64-bit */

static void dc_callvm_free_ppc64(DCCallVM* in_self)
{
  dcFreeMem(in_self);
}

static void dc_callvm_reset_ppc64(DCCallVM* in_self)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;
  dcVecReset(&self->mVecHead);
  self->mIntRegs   = 0;
  self->mFloatRegs = 0;
}

/* fillup integer register file AND push on stack (for ellipsis) */

static void dc_callvm_argInt_ppc64(DCCallVM* in_self, DCint i)
{
  /* promote to longlong */
  dcArgLongLong(in_self, (DClonglong)i);
}

/** floating-point **/

/* double*/

static void dc_callvm_argDouble_ppc64(DCCallVM* in_self, DCdouble d)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;

  if (self->mFloatRegs < 13) {
    self->mRegData.mFloatData[self->mFloatRegs++] = d;
    if (self->mIntRegs < 8) {
      self->mRegData.mIntData[self->mIntRegs++] = *( (DClonglong*) &d );
#if DC__ABI_PPC64_ELF_V == 2
      return;
#endif
    }
  }

#if DC__ABI_PPC64_ELF_V == 2
  if (dcVecSize(&self->mVecHead) == 0) {
    dcVecSkip(&self->mVecHead,sizeof(DClonglong)*8);
  }
#endif

  /* push on stack */
  dcVecAppend(&self->mVecHead,(DCpointer) &d,sizeof(DCdouble));
}

#if DC__ABI_PPC64_ELF_V == 2
static void dc_callvm_argDouble_ppc64_ellipsis(DCCallVM* in_self, DCdouble d)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;

  if (dcVecSize(&self->mVecHead) == 0)
    dcVecSkip(&self->mVecHead,(sizeof(DClonglong))*(self->mIntRegs));

  if (self->mFloatRegs < 13) {
    self->mRegData.mFloatData[self->mFloatRegs++] = d;
    if (self->mIntRegs < 8) {
      self->mRegData.mIntData[self->mIntRegs++] = *( (DClonglong*) &d );
    }
  }

  /* push on stack */
  dcVecAppend(&self->mVecHead,(DCpointer) &d,sizeof(DCdouble));
}
#endif


/* Floating-point */

static void dc_callvm_argFloat_ppc64(DCCallVM* in_self, DCfloat f)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;

  DCdouble d;
#if defined(DC__Endian_BIG)
  struct { DCfloat f_pad; DCfloat f; } sf;
#else /* Endian_LITTLE */
  struct { DCfloat f; DCfloat f_pad; } sf;
#endif

  if (self->mFloatRegs < 13) {
    d = (DCdouble)f;
    self->mRegData.mFloatData[self->mFloatRegs++] = d;
    if (self->mIntRegs < 8) {
      self->mRegData.mIntData[self->mIntRegs++] = *( (DClonglong*) &d );
#if DC__ABI_PPC64_ELF_V == 2
      return;
#endif
    }
  }

#if DC__ABI_PPC64_ELF_V == 2
  if (dcVecSize(&self->mVecHead) == 0) {
    dcVecSkip(&self->mVecHead,sizeof(DClonglong)*8);
  }
#endif

  /* push on stack */
  sf.f = f;
  dcVecAppend(&self->mVecHead,(DCpointer) &sf,sizeof(DCdouble));
}

#if DC__ABI_PPC64_ELF_V == 2
static void dc_callvm_argFloat_ppc64_ellipsis(DCCallVM* in_self, DCfloat f)
{
  /* promote to double */
  dcArgDouble(in_self, (DCdouble) f );
}
#endif


/* long long integer */

static void dc_callvm_argLongLong_ppc64(DCCallVM* in_self, DClonglong L)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;

  /* fillup integer register file */
  if (self->mIntRegs < 8) {
    self->mRegData.mIntData[self->mIntRegs++] = L;
#if DC__ABI_PPC64_ELF_V == 2
    return;
#endif
  }

#if DC__ABI_PPC64_ELF_V == 2
  if (dcVecSize(&self->mVecHead) == 0) {
    dcVecSkip(&self->mVecHead,sizeof(DClonglong)*8);
  }
#endif

  /* push on stack */
  dcVecAppend(&self->mVecHead,&L,sizeof(DClonglong));
}

#if DC__ABI_PPC64_ELF_V == 2
static void dc_callvm_argLongLong_ppc64_ellipsis(DCCallVM* in_self, DClonglong L)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)in_self;

  if (dcVecSize(&self->mVecHead) == 0) 
    dcVecSkip(&self->mVecHead,(sizeof(DClonglong))*(self->mIntRegs));

  if (self->mIntRegs < 8) 
    self->mRegData.mIntData[self->mIntRegs++] = L;

  /* push on stack */
  dcVecAppend(&self->mVecHead,&L,sizeof(DClonglong));
}
#endif


static void dc_callvm_argBool_ppc64(DCCallVM* in_self, DCbool x)
{
  /* promote to longlong */
  dcArgLongLong(in_self, (DClonglong) x );
}


static void dc_callvm_argChar_ppc64(DCCallVM* in_self, DCchar ch)
{
  /* promote to longlong */
  dcArgLongLong(in_self, (DClonglong) ch );
}


static void dc_callvm_argShort_ppc64(DCCallVM* in_self, DCshort s)
{
  /* promote to longlong */
  dcArgLongLong(in_self, (DClonglong) s );
}


static void dc_callvm_argLong_ppc64(DCCallVM* in_self, DClong l)
{
  /* promote to longlong */
  dcArgLongLong(in_self, (DClonglong) l );
}


static void dc_callvm_argPointer_ppc64(DCCallVM* in_self, DCpointer p)
{
  /* promote to longlong */
  dcArgLongLong(in_self, *(DClonglong *) &p );
}


void dc_callvm_call_ppc64(DCCallVM* in_self, DCpointer target)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*) in_self;
  int size = dcVecSize(&self->mVecHead);

  if (size < 64) {
	dcVecSkip(&self->mVecHead, 64-size);
  }
  
  dcCall_ppc64( target, &self->mRegData, dcVecSize(&self->mVecHead) , dcVecData(&self->mVecHead));
}

void dc_callvm_call_ppc64_syscall(DCCallVM* in_self, DCpointer target)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*) in_self;
  dcCall_ppc64_syscall( target, &self->mRegData, dcVecSize(&self->mVecHead) , dcVecData(&self->mVecHead));
}

void dc_callvm_mode_ppc64(DCCallVM* in_self, DCint mode);

DCCallVM_vt gVT_ppc64 =
{
  &dc_callvm_free_ppc64
, &dc_callvm_reset_ppc64
, &dc_callvm_mode_ppc64
, &dc_callvm_argBool_ppc64
, &dc_callvm_argChar_ppc64
, &dc_callvm_argShort_ppc64
, &dc_callvm_argInt_ppc64
, &dc_callvm_argLong_ppc64
, &dc_callvm_argLongLong_ppc64
, &dc_callvm_argFloat_ppc64
, &dc_callvm_argDouble_ppc64
, &dc_callvm_argPointer_ppc64
, NULL /* argStruct */
, (DCvoidvmfunc*)       &dc_callvm_call_ppc64
, (DCboolvmfunc*)       &dc_callvm_call_ppc64
, (DCcharvmfunc*)       &dc_callvm_call_ppc64
, (DCshortvmfunc*)      &dc_callvm_call_ppc64
, (DCintvmfunc*)        &dc_callvm_call_ppc64
, (DClongvmfunc*)       &dc_callvm_call_ppc64
, (DClonglongvmfunc*)   &dc_callvm_call_ppc64
, (DCfloatvmfunc*)      &dc_callvm_call_ppc64
, (DCdoublevmfunc*)     &dc_callvm_call_ppc64
, (DCpointervmfunc*)    &dc_callvm_call_ppc64
, NULL /* callStruct */
};

#if DC__ABI_PPC64_ELF_V == 2
DCCallVM_vt gVT_ppc64_ellipsis =
{
  &dc_callvm_free_ppc64
, &dc_callvm_reset_ppc64
, &dc_callvm_mode_ppc64
, &dc_callvm_argBool_ppc64
, &dc_callvm_argChar_ppc64
, &dc_callvm_argShort_ppc64
, &dc_callvm_argInt_ppc64
, &dc_callvm_argLong_ppc64
, &dc_callvm_argLongLong_ppc64_ellipsis
, &dc_callvm_argFloat_ppc64_ellipsis
, &dc_callvm_argDouble_ppc64_ellipsis
, &dc_callvm_argPointer_ppc64
, NULL /* argStruct */
, (DCvoidvmfunc*)       &dc_callvm_call_ppc64
, (DCboolvmfunc*)       &dc_callvm_call_ppc64
, (DCcharvmfunc*)       &dc_callvm_call_ppc64
, (DCshortvmfunc*)      &dc_callvm_call_ppc64
, (DCintvmfunc*)        &dc_callvm_call_ppc64
, (DClongvmfunc*)       &dc_callvm_call_ppc64
, (DClonglongvmfunc*)   &dc_callvm_call_ppc64
, (DCfloatvmfunc*)      &dc_callvm_call_ppc64
, (DCdoublevmfunc*)     &dc_callvm_call_ppc64
, (DCpointervmfunc*)    &dc_callvm_call_ppc64
, NULL /* callStruct */
};
#endif

DCCallVM_vt gVT_ppc64_syscall =
{
  &dc_callvm_free_ppc64
, &dc_callvm_reset_ppc64
, &dc_callvm_mode_ppc64
, &dc_callvm_argBool_ppc64
, &dc_callvm_argChar_ppc64
, &dc_callvm_argShort_ppc64
, &dc_callvm_argInt_ppc64
, &dc_callvm_argLong_ppc64
, &dc_callvm_argLongLong_ppc64
, &dc_callvm_argFloat_ppc64
, &dc_callvm_argDouble_ppc64
, &dc_callvm_argPointer_ppc64
, NULL /* argStruct */
, (DCvoidvmfunc*)       &dc_callvm_call_ppc64_syscall
, (DCboolvmfunc*)       &dc_callvm_call_ppc64_syscall
, (DCcharvmfunc*)       &dc_callvm_call_ppc64_syscall
, (DCshortvmfunc*)      &dc_callvm_call_ppc64_syscall
, (DCintvmfunc*)        &dc_callvm_call_ppc64_syscall
, (DClongvmfunc*)       &dc_callvm_call_ppc64_syscall
, (DClonglongvmfunc*)   &dc_callvm_call_ppc64_syscall
, (DCfloatvmfunc*)      &dc_callvm_call_ppc64_syscall
, (DCdoublevmfunc*)     &dc_callvm_call_ppc64_syscall
, (DCpointervmfunc*)    &dc_callvm_call_ppc64_syscall
, NULL /* callStruct */
};

void dc_callvm_mode_ppc64(DCCallVM* in_self, DCint mode)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*) in_self;
  DCCallVM_vt* vt;
  switch(mode) {

    case DC_CALL_C_PPC64: 
    case DC_CALL_C_DEFAULT:
    case DC_CALL_C_ELLIPSIS:
#if DC__ABI_PPC64_ELF_V == 2
      vt = &gVT_ppc64;
      break;
    case DC_CALL_C_ELLIPSIS_VARARGS:
      vt = &gVT_ppc64_ellipsis;
      break;
#else
    case DC_CALL_C_ELLIPSIS_VARARGS:
      vt = &gVT_ppc64;
      break;
#endif

    case DC_CALL_SYS_DEFAULT:
    case DC_CALL_SYS_PPC64:
      vt = &gVT_ppc64_syscall;
      break;

    default: 
      self->mInterface.mError = DC_ERROR_UNSUPPORTED_MODE; 
      return;
  }
  
  dc_callvm_base_init(&self->mInterface, vt);
}

DCCallVM* dcNewCallVM(DCsize size)
{
  DCCallVM_ppc64* self = (DCCallVM_ppc64*)dcAllocMem(sizeof(DCCallVM_ppc64)+size);
  dcVecInit(&self->mVecHead, size);
  self->mIntRegs              = 0;
  self->mFloatRegs            = 0;
  dc_callvm_mode_ppc64( (DCCallVM*) self, DC_CALL_C_DEFAULT );
  return (DCCallVM*)self;
}

